rust/src/instruction/mod.rs

```diff
@@ -0,0 +1,217 @@
+use pyo3::{
+    exceptions::{PyRuntimeError, PyValueError},
+    prelude::*,
+    types::PyBytes,
+};
+
+use std::{collections::HashMap, str::FromStr};
+
+use quil_rs::quil::Quil;
+
+mod declaration;
+mod gate;
+mod quilt;
+
+pub use declaration::Declare;
+pub use gate::*;
+pub use quilt::*;
+
+use crate::impl_eq;
+
+pub fn init_module(_py: Python<'_>) -> PyResult<Bound<'_, PyModule>> {
+    let m = PyModule::new_bound(_py, "instruction")?;
+    m.add_class::<Instruction>()?;
+    m.add_class::<DefCalibration>()?;
+    m.add_class::<DefMeasureCalibration>()?;
+    m.add_class::<GateModifier>()?;
+    Ok(m)
+}
+
+#[pyclass(subclass)]
+#[derive(Clone, Debug, PartialEq)]
+pub struct Instruction {
+    inner: quil_rs::instruction::Instruction,
+}
+impl_eq!(Instruction);
+
+impl From<Instruction> for quil_rs::instruction::Instruction {
+    fn from(instruction: Instruction) -> Self {
+        instruction.inner
+    }
+}
+
+impl From<quil_rs::instruction::Instruction> for Instruction {
+    fn from(instruction: quil_rs::instruction::Instruction) -> Self {
+        Self { inner: instruction }
+    }
+}
+
+#[pymethods]
+impl Instruction {
+    fn out(&self) -> PyResult<String> {
+        self.inner
+            .to_quil()
+            .map_err(|e| PyValueError::new_err(e.to_string()))
+    }
+
+    pub fn __str__(&self) -> String {
+        self.inner.to_quil_or_debug()
+    }
```

Comment:
Wondering if it makes sense to have `__str__` error if the instruction is not valid Quil

Reply:
Not for pyQuil. Since v3 (at least) `out()` has been the method that returns valid Quil or raises an exception and `__str__` would do its best.